core/chains/evm/headtracker/head_saver_in_mem.go

```diff
@@ -0,0 +1,163 @@
+package headtracker
+
+import (
+	"context"
+	"math/big"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+
+	htrktypes "github.com/smartcontractkit/chainlink/v2/common/headtracker/types"
+	"github.com/smartcontractkit/chainlink/v2/common/types"
+	evmtypes "github.com/smartcontractkit/chainlink/v2/core/chains/evm/types"
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+)
+
+type inMemoryHeadSaver[H types.HeadTrackerHead[BLOCK_HASH, CHAIN_ID], BLOCK_HASH types.Hashable, CHAIN_ID types.ID] struct {
+	config      htrktypes.Config
+	logger      logger.Logger
+	latestHead  H
+	Heads       map[BLOCK_HASH]H
+	HeadsNumber map[int64][]H
+	mu          sync.RWMutex
+	getNilHead  func() H
+	setParent   func(H, H)
+}
+
+type EvmInMemoryHeadSaver = inMemoryHeadSaver[*evmtypes.Head, common.Hash, *big.Int]
+
+var _ types.HeadSaver[*evmtypes.Head, common.Hash] = (*EvmInMemoryHeadSaver)(nil)
+
+func NewInMemoryHeadSaver[
+	H types.HeadTrackerHead[BLOCK_HASH, CHAIN_ID],
+	BLOCK_HASH types.Hashable,
+	CHAIN_ID types.ID](
+	config htrktypes.Config,
+	lggr logger.Logger,
+	getNilHead func() H,
+	setParent func(H, H),
+) *inMemoryHeadSaver[H, BLOCK_HASH, CHAIN_ID] {
+	return &inMemoryHeadSaver[H, BLOCK_HASH, CHAIN_ID]{
+		config:      config,
+		logger:      lggr.Named("InMemoryHeadSaver"),
+		Heads:       make(map[BLOCK_HASH]H),
+		HeadsNumber: make(map[int64][]H),
+		getNilHead:  getNilHead,
+		setParent:   setParent,
+	}
+}
+
+func NewEvmInMemoryHeadSaver(config Config, lggr logger.Logger) *EvmInMemoryHeadSaver {
+	evmConfig := NewWrappedConfig(config)
+	return NewInMemoryHeadSaver[*evmtypes.Head, common.Hash, *big.Int](
+		evmConfig,
+		lggr,
+		func() *evmtypes.Head { return nil },
+		func(head, parent *evmtypes.Head) { head.Parent = parent },
+	)
+}
+
+func (hs *inMemoryHeadSaver[H, BLOCK_HASH, CHAIN_ID]) Save(ctx context.Context, head H) error {
+	if !head.IsValid() {
+		return nil
```

Comment:
should be return error here

Reply:
thanks for pointing this out, my mistake for not handling the error here.